_posts/2020-08-05-using-open-policy-agent-with-strimzi-and-apache-kafka.md

```diff
@@ -0,0 +1,456 @@
+
+The `keycloak` authorization is based on the [Keycloak Authorization Services](https://www.keycloak.org/docs/latest/authorization_services/index.html).
+It can be used only in combination with OAuth 2.0 authentication.
+Security policies and permissions are defined in Keycloak and are used to grant access to resources on Kafka brokers. 
+Users and clients are matched against policies that permit access to perform specific actions on Kafka brokers.
+You can also use the _federation_ feature of Keycloak to map the policies for example against LDAP groups.
+The `KeycloakRBACAuthorizer` will fetch the list of granted permissions from Keycloak and enforce the authorization rules locally to make sure it doesn't negatively impact performance.
+
+The main advantage of Keycloak is that you can have everything managed centrally.
+Not just for Strimzi and Apache Kafka, but for all your applications which support OAuth.
+It can also handle both authentication and authorization.
+So you do not need to have different servers for authentication and authorization and all data are stored in one place.
+
+On the other hand, you will still need to run and manage Keycloak.
+So unless you already use it or at least plan to use some central authentication / authorization server anyway, this will be additional effort you would need to deal with.
+
+## Open Policy Agent
+
+[Open Policy Agent](https://www.openpolicyagent.org/) (OPA) is an open source policy engine.
+It can be used to enforce policies across your whole architecture stack.
+The policies are declarative and to write them it is using its own language called _Rego_.
+OPA can be used for policy based control of many different applications.
+It can be used with Kubernetes, APIs, SSH, CI/CD pipelines and many more.
+
+![Open Policy Agent](/assets/images/posts/2020-08-05-using-open-policy-agent-with-strimzi-and-apache-kafka.png)
+
+OPA decouples the policy decision making from enforcing the decision.
+When used with Kafka, the authorizer running inside Kafka will call OPA server to evaluate the policy based on the input from the authorizer.
+The input will be the same set of information as with any other Kafka authorizer.
+It is described in the introduction to this blog post.
+OPA will evaluate the policy and respond to the authorizer request with a decision.
+And the authorizer will either allow or deny the operation.
+The decisions are of course cached by the authorizer to make sure the performance of the Kafka cluster is not affected.
+
+We didn't developed our own OPA Authorizer for Kafka.
+Instead we are using the existing [OPA Plugin from Bisnode](https://github.com/Bisnode/opa-kafka-plugin).
+
+### Configuring OPA authorizer
+
+Before we configure our Kafka cluster to use OPA, we first need to deploy the OPA server.
+If you don't use OPA yet, you can follow [this guide](https://www.openpolicyagent.org/docs/latest/deployments/#kubernetes) from the OPA documentation.
```

Comment:
```suggestion
If you do not have an OPA deployed yet, you can follow [this guide](https://www.openpolicyagent.org/docs/latest/deployments/#kubernetes) the OPA deployment documentation.
```

Reply:
I think the `from` was supposed to stay there? Also the `an` in `an OPA` sounds very weird to me. Should it really be there?

Reply:
Sorry a typo on my part.